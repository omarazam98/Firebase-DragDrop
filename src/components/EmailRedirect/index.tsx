import React from 'react';
import { withAPI } from '@winwin/api-firebase';

export const EmailRedirect = (props) => {
  const handleClick = () => props.history.push('/login');
  const handleSendClick = () => {
    const redirect = window['_env_'].NODE_ENV === 'production' ? 'http://www.winwinhomesharing.com'
      : window['_env_'].REACT_APP_EMAIL_REDIRECT || 'http://integration.scoutmastersforever.com';
    props.api.auth.currentUser().sendEmailVerification({ url: redirect });
  };

  return (
    <div>
      Email has been sent to verify account. <br/>
      Already verified?
      <button onClick={handleClick} id="redirect"> Log In</button> <br/>
      Didn't receive the email?
      <button onClick={handleSendClick} id="resend"> Send Again</button>
    </div>
  );
};
export default withAPI(EmailRedirect);
