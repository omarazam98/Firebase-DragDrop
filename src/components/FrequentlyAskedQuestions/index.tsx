import React, { useEffect, useState } from 'react';
import { withAPI } from '@winwin/api-firebase';
import FrequentQuestions from './FrequentQuestions';

interface FrequentlyAskedQuestionsPageProps {
  api: any;
}

const FrequentlyAskedQuestionsPage: React.FC<FrequentlyAskedQuestionsPageProps> = (props) => {
  const [questionList, setQuestionList] = useState([]);

  useEffect(() => {
    const frequentQuestionsCollection = props.api.data.faq.getAll();
    frequentQuestionsCollection.then(questions =>
      setQuestionList(questionList => questions));
  },        [props.api.data.faq]);

  const handleSubmit = function (question:string, email?:string) {
    props.api.data.faq.addQuestion(question, email);
  };

  return (
    <div>
      <FrequentQuestions questionList={questionList} handleSubmit={handleSubmit}/>
    </div>
  );

};

export default withAPI(FrequentlyAskedQuestionsPage);
