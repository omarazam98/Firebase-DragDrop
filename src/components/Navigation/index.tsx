// Main routing component.
// If a new page is added, be sure to add the route to src/constants/routes
import React from 'react';
import { BrowserRouter as Router, Route, Link } from 'react-router-dom';
import Routes from '../../constants/routes';
import {withAPI} from '@winwin/api-firebase';
import SplitPane from 'react-split-pane';

export function Navigation(props) {
  return (
    <div>
      <Router>
        <SplitPane split="vertical" minSize={50} defaultSize={100}>
          {/* Link creates the object that a user can click on to go to another page */}
          <div>
            {links()}
            <button onClick = {() => {props.api.auth.signOut()}}>Log Out</button>
          </div>
          {/* Route indicates what component should be shown, based on what is linked */}
          {/* Map routes from links to their components */}
          <div>
          {Routes.map((route:any, index) => (
            <Route
              key={index}
              path={route.path}
              exact={route.exact}
              component={route.component}
            />
          ))}
          </div>
        </SplitPane>
      </Router>
    </div>
  );
}
// iterate over all routes from routes.ts
// return a jsx expression containing links to all routes
function links() {
  // can't push JSX on an empty array, so init with a div
  const allLinks = [<div />];
  Routes.forEach((route:any) => allLinks.push(<Link to={route.path}>{route.name}<br /></Link>));
  return allLinks;
}

export default withAPI(Navigation);
