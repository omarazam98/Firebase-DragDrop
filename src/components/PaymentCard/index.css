.credit-card {
    display: flex;
    flex-direction: column;
    justify-content: space-between;

    width: 50vw;
    height: 32vw;
    padding: 25px;
    border-radius: 15px;
    max-width: 400px;
    max-height: 250px;
    color: white;

    background-image: linear-gradient(25deg, #0f509e, #1399cd);
    box-shadow: 0 0 20px rgba(0, 0, 0, 0.5);
}

.credit-card__number {
    display: flex;
    justify-content: center;

    font-family: 'Arial', sans-serif;
    font-size: 2em;
}

.credit-card__info {
    display: flex;
    justify-content: space-between;

    font-family: 'Arial', sans-serif;
    font-size: 2em;
}

.credit-card__info_label {
    font-size: 0.4em;
}

.credit-card__info_expiry {
    display: flex;
    flex-direction: column;
    align-items: flex-end;
}
