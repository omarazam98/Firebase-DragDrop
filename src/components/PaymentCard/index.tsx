import React from 'react';
import './index.css';
import { withAPI } from '@winwin/api-firebase';

interface PaymentMethods {
  type: string;
  number: number;
}

interface UserType {
  displayName: string;
  uid: string;
}
interface PaymentProps {
  api: any;
}
interface PaymentState {
  paymentMethod: PaymentMethods | undefined;
  user: UserType | undefined;
}
const INITIAL_STATE = {
  paymentMethod: undefined,
  user: undefined,
};

export class PaymentCard extends React.Component<PaymentProps, PaymentState>{
  constructor(props) {
    super(props);
    this.state = INITIAL_STATE;
  }

  componentDidMount() {
    const user = this.props.api.auth.currentUser();
    if (user) {
      this.props.api.data.users.getPaymentMethods(user.uid).then((payments) => {
        this.setState(() => {
          return {
            user,
            paymentMethod: payments,
          };
        });
      });
    }
  }

  render() {
    return(
      <div className="credit-card">
        <div className="credit-card__logo">Win-Win Payment Method</div>
        <div className="credit-card__number">
          {this.state.paymentMethod && `XXXX XXXX XXXX ${this.state.paymentMethod.number}`}
        </div>

        <div className="credit-card__info">
          <div className="credit-card__info_name">
            <div className="credit-card__info_label">CARDHOLDER'S NAME</div>
            <div>{this.state.user && this.state.user.displayName}</div>
          </div>

          <div className="credit-card__info_expiry">
            <button>Change</button>
          </div>
        </div>

      </div>
    );
  }
}

export default withAPI(PaymentCard);
