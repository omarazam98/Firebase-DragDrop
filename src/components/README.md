## What are components?
Components are reusable, independent pieces of user interface.

They are a core concept in react, so if this is news to you check out the docs at:
https://reactjs.org/docs/components-and-props.html 

## How do we use components?
Components can be completely independent vanilla react components. In that case data is passed to them by their parent with props, and is modified locally or by their children using this.state. These are generally called presentational components, and have no concept of redux.

However if components to interact with data from either Redux, or the database they do so in two ways.

1. Receiving data from redux through mapStateToProps
2. Dispatching action to redux through mapDispatchToProps

Components that either need to receive data from redux or dispatch actions to redux need to be connected to the redux store. This is done as follows.

**If any of this is confusing this resource is quite good https://blog.logrocket.com/react-redux-connect-when-and-how-to-use-it-f2a1edab2013/**

For this example assume that this component has been called as a decendant of <Provider> 
We have a redux store that has this structure describing a garage:

  {
    cars: [
      {},
      {},
      {}
    ],
    open: false
  }

And an action creator that looks like this

  const toggleDoor = () => {
    return {
      type: 'TOGGLE_DOOR',
    }
  }

The component would then look like this

  import React, { Component } from 'react';
  import { connect } from 'react-redux';
  import { changeName } from '../actions/fileThatHasThisCreatorInIt'

  // This class should be exported as a named export to be used in testing.
  // However as explained at the end of this example this export will not be the one used in source code
  export class SomeComponent extends Component {
    constructor(props){
      super(props)

      // if you are unfamiliar with binding check this out https://reactjs.org/docs/handling-events.html
      this.openCloseGarage = this.openCloseGarage.bind(this);
    }

    openCloseGarage() {
      // toggleGarageDoor is available as a prop, because we have used mapDispatchToProps seen below
      this.props.toggleGarageDoor();
    }

    render() {
      return (
        <div>
          <button onClick={this.openCloseGarage}>
            // garageOpen is available as a prop beause we provided it through mapStateToProps seen below
            {this.props.garageOpen}
          </button>
        </div>
      )
    }

    // This is where the redux specific details are added

    // mapStateToProps is used to subscribe to data from the redux store.
    // It allows data from the store to be accessed as a prop in the component
    // The first variable is the redux store state, and the second are the props that are provided by the component constructor (Normal props not mapped to redux state)

    const mapStateToProps = (reduxState, currentProps) => {
      return {
        // this is where the prop garageOpen is mapped to the 'open' field of theh state of the redux store
        garageOpen: state.open,
      }
    }

    // mapDispatchToProps is used to gain access to dispatching actions as a component prop
    // The first variable is the store.dispatch command, and the second are the props that are providd by the component constructor (Normal props not mapped to redux state)

    const mapDispatchToProps = (dispatch, currentProps) => {
      return {
        // this is where the prop toggleGarageDoor is mapped to the dispatching of the action toggleDoor
        // For this one ensure that you are declaring an inline function that calls dispatch with the action creator as the argument
        toggleGarageDoor: () => { dispatch(toggleDoor()) }
      }
    }

    // Finally the functions that were defined above need to be connected to the component.
    // Ensure that this connected component is what is exported as default and used in the source code.
    // However it is useful to export the unconnected component as a named export for testing

    connect(mapStateToProps, mapDispatchToProps)(SomeComponent)

  }