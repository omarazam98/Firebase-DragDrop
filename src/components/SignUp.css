#phoneNumber {
    background-color: orange
}

#email {
    background-color: red

}

#passwordOne {
    background-color: green;

}

#passwordTwo {
    background-color: blue

}

#phoneNumber p.ws-errormessage::after {
    content:"Please enter the phone number in the form (xxx) xxx-xxxx";
}

.error { border: 1px solid red; }

.form-control:valid~.invalid-feedback {display: none;}
