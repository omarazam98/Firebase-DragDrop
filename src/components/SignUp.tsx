import React, { Component } from 'react';
import './SignUp.css';
import { fireauth, firestore} from '../firebase/firebase'

interface SignUpState {
    name: {
        value: string,
        touched: boolean,
        errors: string,
    };
    phoneNumber: {
        value: string,
        touched: boolean,
        errors: string,
    };
    class:{
        value: string,
        touched: boolean,
        errors: string,
    };
    email: {
        value: string,
        touched: boolean,
        errors: string,
    };
    passwordOne: {
        value: string,
        touched: boolean,
        errors: string,
    };
    passwordTwo: {
        value: string,
        touched: boolean,
        errors: string,
    };
}

const INITIAL_STATE: SignUpState = {
    name: {
        value: '',
        touched: false,
        errors: '',
    },
    phoneNumber: {
        value: '',
        touched: false,
        errors: '',
    },
    class: {
        value: '',
        touched: false,
        errors: '',
    },
    email: {
        value: '',
        touched: false,
        errors: '',
    },
    passwordOne: {
        value: '',
        touched: false,
        errors: '',
    },
    passwordTwo: {
        value: '',
        touched: false,
        errors: '',
    },
};

class SignUp extends Component<any, SignUpState> {
    constructor(props) {
        super(props);
        this.state = INITIAL_STATE;
    }

    onChange = e => {
        e.persist()

        if (Object.keys(this.state).includes(e.target.name)) {
            this.setState({
                [e.target.name]: {
                    ...this.state[e.target.name],
                    value: e.target.value
                }
            } as Pick<SignUpState, keyof SignUpState>, () => {
                this.validateInputs(e);
            });

        };
    };
    validateInputs = e => {
        if (e.target.value === '') {
            this.setState({
                [e.target.name]: {
                    ...this.state[e.target.name],
                    errors: 'field cannot be empty'
                }
            } as Pick<SignUpState, keyof SignUpState>);
        }else if (e.target.name === 'passwordTwo' && this.state.passwordOne.value !== this.state.passwordTwo.value) {
            this.setState({
                passwordTwo: {
                    ...this.state.passwordTwo,
                    errors: 'passwords must match'
                }
            } as Pick<SignUpState, keyof SignUpState>);
        } else if (!e.target.checkValidity()) {
            this.setState({
                [e.target.name]: {
                    ...this.state[e.target.name],
                    errors: 'invalid format'
                }
            } as Pick<SignUpState, keyof SignUpState>);
        } else {
            this.setState({
                [e.target.name]: {
                    ...this.state[e.target.name],
                    errors: ''
                }
            } as Pick<SignUpState, keyof SignUpState>);
        };
    };
    submit = e => {
        fireauth.createUserWithEmailAndPassword(this.state.email.value, this.state.passwordOne.value).then((userCredential) => {
            const user = userCredential.user;
            if (user){
                user.updateProfile({ displayName: this.state.name.value }).then(()=>{
                    console.log(user.uid);
                    console.log(user.emailVerified);
                    firestore.collection('users').doc(user.uid).set({email: user.email, phoneNumber: this.state.phoneNumber.value, name: user.displayName, class: this.state.class.value});
                });
            }
        }).catch(function(error) {
            var errorCode = error.code;
            var errorMessage = error.message;
            if (errorCode == 'auth/weak-password') {
                alert('The password is too weak.');
            } else {
                alert(errorMessage);
            }
        });
        e.preventDefault();
    }

    hasErrors = () => {
        for(const key in this.state){
            if (this.state[key].errors || this.state[key].value === ''){
                return true;
            }
        }
        return false;
    };

    handleBlur = e => {
        if (Object.keys(this.state).includes(e.target.name)) {
            this.setState({[e.target.name]: { ...this.state[e.target.name], touched: true}} as Pick<SignUpState, keyof SignUpState>);
        };
    };

    render(){
        const disableButton = this.hasErrors();
        return(
            <div>
                <h1> Sign Up Page </h1>
                <form onSubmit={this.submit}>
                    <label htmlFor="name">Name: </label> <br/>
                    <input name="name" id="name" onBlur = {this.handleBlur} value={this.state.name.value} onChange={this.onChange} type="text" placeholder="Full Name"/> {this.state.name.touched && <span> {this.state.name.errors} </span> } <br/>

                    <input type="radio" name="class" value="senior" checked={this.state.class.value === "senior"} onChange={this.onChange} />Senior
                    <input type="radio" name="class" value="student" checked={this.state.class.value === "student"} onChange={this.onChange} />Student <br/>

                    <label htmlFor="email">Email: </label> <br/>
                    <input name="email" id="email" onBlur = {this.handleBlur} value={this.state.email.value} onChange={this.onChange} type="email" placeholder="Email Address"/>  {this.state.email.touched && <span> {this.state.email.errors} </span> }<br/>
                    <label htmlFor="phoneNumber">Phone Number: </label><br/>
                    <input name="phoneNumber" id='phoneNumber' onBlur={this.handleBlur} onChange={this.onChange} type="tel" placeholder="Phone Number" pattern="[0-9]{3}-[0-9]{3}-[0-9]{4}"/>{this.state.phoneNumber.touched && <span> {this.state.phoneNumber.errors} </span> }<br/>
                    <label htmlFor="passwordOne">Password: </label><br/>
                    <input name="passwordOne" id="passwordOne" onBlur={this.handleBlur} value={this.state.passwordOne.value} onChange={this.onChange} type="password" placeholder="Password" />{this.state.passwordOne.touched && <span> {this.state.passwordOne.errors} </span> } <br/>
                    <label htmlFor="passwordTwo">Confirm Password: </label> <br/>
                    <input name="passwordTwo" id="passwordTwo"onBlur={this.handleBlur} value={this.state.passwordTwo.value} onChange={this.onChange} type="password" placeholder="Confirm Password" />{ this.state.passwordTwo.touched && <span> {this.state.passwordTwo.errors} </span> } <br/>
                    <input disabled={disableButton} type="submit" value={'Sign Up'}></input>
        </form>
            </div>
        );
    };
}
export default SignUp;
