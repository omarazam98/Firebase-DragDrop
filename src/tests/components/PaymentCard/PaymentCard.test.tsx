import * as React from 'react';
import { PaymentCard, styles } from '../../../components/PaymentCard';
import { shallow } from 'enzyme';

const mockState = {
  paymentMethod: {
    type: 'visa',
    number: 1010,
  },
  user: {
    uid: '1111',
    displayName: 'The Weapon',
  },
};

const mockAPI = {
  auth: {
    currentUser: () => {
      jest.fn();
    },
  },
};

test('Render the payment card with no data', () => {
  const wrapper = shallow(<PaymentCard classes={styles} api={mockAPI}/>);
  expect(wrapper).toMatchSnapshot();
});

test('Render full payment card with data', () => {
  const wrapper = shallow(<PaymentCard classes={styles} api={mockAPI}/>);
  wrapper.setState(mockState);
  expect(wrapper).toMatchSnapshot();
  expect(wrapper.find('#userName').text())
    .toBe(mockState.user.displayName);
  expect(wrapper.find('#payment').text())
    .toBe(`XXXX XXXX XXXX ${mockState.paymentMethod.number}`);
});
