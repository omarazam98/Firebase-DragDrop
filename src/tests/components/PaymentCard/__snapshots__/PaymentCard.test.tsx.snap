// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render full payment card with data 1`] = `
<div>
  <WithStyles(ForwardRef(Card))
    className={
      Object {
        "backgroundImage": "linear-gradient(25deg, #0f509e, #1399cd)",
        "borderRadius": "15px",
        "boxShadow": "0 0 20px rgba(0, 0, 0, 0.5)",
        "color": "white",
        "height": "32vw",
        "maxHeight": "250px",
        "maxWidth": "400px",
        "padding": "25px",
        "position": "relative",
        "width": "50vw",
      }
    }
  >
    <WithStyles(ForwardRef(CardContent))>
      <WithStyles(ForwardRef(Typography))
        className={
          Object {
            "fontSize": 14,
          }
        }
        color="textPrimary"
      >
        Win-Win Payment Method
      </WithStyles(ForwardRef(Typography))>
      <WithStyles(ForwardRef(Typography))
        className={
          Object {
            "display": "flex",
            "fontFamily": "\\"Arial\\", sans-serif",
            "fontSize": "2em",
            "position": "absolute",
            "top": "40%",
          }
        }
        color="textPrimary"
        id="payment"
      >
        XXXX XXXX XXXX 1010
      </WithStyles(ForwardRef(Typography))>
      <div
        className={
          Object {
            "bottom": "1vw",
            "position": "absolute",
          }
        }
      >
        <WithStyles(ForwardRef(Typography))
          className={
            Object {
              "fontSize": "1em",
            }
          }
          color="textPrimary"
        >
          CARDHOLDER'S NAME
        </WithStyles(ForwardRef(Typography))>
        <WithStyles(ForwardRef(Typography))
          className={
            Object {
              "display": "flex",
              "fontFamily": "\\"Arial\\", sans-serif",
              "fontSize": "2em",
              "justifyContent": "space-between",
            }
          }
          color="textPrimary"
          id="userName"
        >
          The Weapon
        </WithStyles(ForwardRef(Typography))>
      </div>
    </WithStyles(ForwardRef(CardContent))>
    <WithStyles(ForwardRef(CardActions))>
      <WithStyles(ForwardRef(Button))
        className={
          Object {
            "bottom": "1vw",
            "position": "absolute",
            "right": "1vw",
            "variant": "contained",
          }
        }
        variant="contained"
      >
        Change
      </WithStyles(ForwardRef(Button))>
    </WithStyles(ForwardRef(CardActions))>
  </WithStyles(ForwardRef(Card))>
</div>
`;

exports[`Render the payment card with no data 1`] = `
<div>
  <WithStyles(ForwardRef(Card))
    className={
      Object {
        "backgroundImage": "linear-gradient(25deg, #0f509e, #1399cd)",
        "borderRadius": "15px",
        "boxShadow": "0 0 20px rgba(0, 0, 0, 0.5)",
        "color": "white",
        "height": "32vw",
        "maxHeight": "250px",
        "maxWidth": "400px",
        "padding": "25px",
        "position": "relative",
        "width": "50vw",
      }
    }
  >
    <WithStyles(ForwardRef(CardContent))>
      <WithStyles(ForwardRef(Typography))
        className={
          Object {
            "fontSize": 14,
          }
        }
        color="textPrimary"
      >
        Win-Win Payment Method
      </WithStyles(ForwardRef(Typography))>
      <WithStyles(ForwardRef(Typography))
        className={
          Object {
            "display": "flex",
            "fontFamily": "\\"Arial\\", sans-serif",
            "fontSize": "2em",
            "position": "absolute",
            "top": "40%",
          }
        }
        color="textPrimary"
        id="payment"
      />
      <div
        className={
          Object {
            "bottom": "1vw",
            "position": "absolute",
          }
        }
      >
        <WithStyles(ForwardRef(Typography))
          className={
            Object {
              "fontSize": "1em",
            }
          }
          color="textPrimary"
        >
          CARDHOLDER'S NAME
        </WithStyles(ForwardRef(Typography))>
        <WithStyles(ForwardRef(Typography))
          className={
            Object {
              "display": "flex",
              "fontFamily": "\\"Arial\\", sans-serif",
              "fontSize": "2em",
              "justifyContent": "space-between",
            }
          }
          color="textPrimary"
          id="userName"
        />
      </div>
    </WithStyles(ForwardRef(CardContent))>
    <WithStyles(ForwardRef(CardActions))>
      <WithStyles(ForwardRef(Button))
        className={
          Object {
            "bottom": "1vw",
            "position": "absolute",
            "right": "1vw",
            "variant": "contained",
          }
        }
        variant="contained"
      >
        Change
      </WithStyles(ForwardRef(Button))>
    </WithStyles(ForwardRef(CardActions))>
  </WithStyles(ForwardRef(Card))>
</div>
`;
